web_poet/rules.py

```diff
@@ -259,6 +317,34 @@ def search_overrides(self, **kwargs) -> List[ApplyRule]:
         warnings.warn(msg, DeprecationWarning, stacklevel=2)
         return self.search(**kwargs)
 
+    def _rules_for_url(
+        self, url: Union[_Url, str], matchers: Dict[Any, URLMatcher]
+    ) -> Mapping[Type, Type[ItemPage]]:
+        result: Dict[Type, Type[ItemPage]] = {}
+
+        url = str(url)
+
+        for target, matcher in matchers.items():
+            rule_id = matcher.match(url)
+            if rule_id is not None:
+                result[target] = self._rules[rule_id].use
+
+        return result
+
+    def overrides_for(self, url: Union[_Url, str]) -> Mapping[Type, Type[ItemPage]]:
+        """Finds all of the page objects associated with the given URL and
+        returns a Mapping where the 'key' represents the page object that is
+        **overridden** by the page object in 'value'.
+        """
+        return self._rules_for_url(url, self._overrides_matchers)
+
+    def page_object_for(self, url: Union[_Url, str]) -> Mapping[Type, Type[ItemPage]]:
+        """Finds all of the page objects associated with the given URL and
+        returns a Mapping where the 'key' represents the item class that is
+        **produced** by the page object in 'value'.
```

Comment:
What's the use case for returning all item classes? If I'm not mistaken, in scrapy-poet the most common case is getting 1 item type. Currently if you need to know which page object handles MyProduct data type for example.com, we'll run matching for example.com for all registered item types, not just for MyProduct.

Basically, _rules_for_url iterates over all targets and runs matching for all of them, but it looks pointless, because we know the target already (the item class); it's a key in the matchers dict, so we can look it up easily.

Reply:
Apart from being symmetric with `overrides_for()`, it offers discoverability on the types of items that can be extracted from a given URL. This leads to several possible functionalities. For example, in **scrapy-poet**, there could be a wildcard which indicates "give me all the items you can extract from this URL based on its pattern". This is useful in scenarios where product discovery leads to various items with different URL patterns. 

Should we drop the current `page_object_for()` for now?

However, you're right about the inefficiencies for `page_object_for()` and I've created a new more efficient `page_object_for_item()` for it.

Reply:
hey! This makes sense; there could be some use cases which require getting everything. But I'd probably drop the method for now, until we face them.

Reply:
Done in https://github.com/scrapinghub/web-poet/pull/112/commits/a5e8480f9d20bf65f6ecc8b0d6e56e666abb1223